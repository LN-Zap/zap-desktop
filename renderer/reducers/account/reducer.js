import createReducer from '@zap/utils/createReducer'
import { getIntl } from '@zap/i18n'
import { mainLog } from '@zap/utils/log'
import waitForIpcEvent from '@zap/utils/waitForIpc'
import { closeDialog } from 'reducers/modal'
import { showNotification } from 'reducers/notification'
import accountSelectors from './selectors'
import messages from './messages'
import * as constants from './constants'

const {
  INIT_ACCOUNT,
  INIT_ACCOUNT_SUCCESS,
  INIT_ACCOUNT_FAILURE,
  LOGIN,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGIN_CLEAR_ERROR,
  CHANGE_PASSWORD_DIALOG_ID,
  PASSWORD_PROMPT_DIALOG_ID,
  PASSWORD_SET_DIALOG_ID,
  SET_IS_PASSWORD_ENABLED,
  LOGIN_NOT_ALLOWED,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

export const initialState = {
  isAccountLoading: false,
  isAccountLoaded: false,
  initAccountError: null,
  isLoggingIn: false,
  loginError: null,
  isLoggedIn: false,
  isPasswordEnabled: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * initAccount - Fetch the current account info from the database and save into the store.
 * Should be called once when the app first loads.
 *
 * @returns {Function} Thunk
 */
export const initAccount = () => async dispatch => {
  dispatch({ type: INIT_ACCOUNT })
  try {
    // Auto login user if password feature is disabled.
    const isAccountPasswordEnabled = await dispatch(checkAccountPasswordEnabled())
    if (!isAccountPasswordEnabled) {
      dispatch({ type: LOGIN_SUCCESS })
    }
    dispatch({ type: INIT_ACCOUNT_SUCCESS })
  } catch (e) {
    mainLog.warn('checkAccountPasswordEnabled error: %o', e)
    dispatch({ type: LOGIN_FAILURE, error: LOGIN_NOT_ALLOWED })
  }
}

/**
 * setIsPasswordEnabled - Whether password is set for the account.
 *
 * @param {string} value value
 * @returns {object} Action
 */
const setIsPasswordEnabled = value => ({
  type: SET_IS_PASSWORD_ENABLED,
  value,
})

/**
 * setPassword - Updates wallet password.
 *
 * @param {string} password new password
 * @returns {Function} Thunk
 */
const setPassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  dispatch(waitForIpcEvent('setPassword', { value: await sha256digest(password) }))
}

/**
 * checkAccountPasswordEnabled - Checks whether app password is set via checking secure storage.
 * Dispatches setIsPasswordEnabled on completion.
 *
 * @returns {Function} Thunk
 */
const checkAccountPasswordEnabled = () => async dispatch => {
  const { value } = await dispatch(waitForIpcEvent('hasPassword'))
  dispatch(setIsPasswordEnabled(value))
  return value
}

/**
 * changePassword - Changes existing password.
 *
 * @param {object} params password params
 * @param {string} params.newPassword new password
 * @param {string} params.oldPassword old password
 * @returns {Function} Thunk
 */
export const changePassword = ({ newPassword, oldPassword }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(oldPassword))
    await dispatch(setPassword(newPassword))
    dispatch(closeDialog(CHANGE_PASSWORD_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_updated)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * enablePassword - Enables app-wide password protection.
 *
 * @param {string} password to be used further on.
 * @returns {Function} Thunk
 */
export const enablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(setPassword(password))
    dispatch(closeDialog(PASSWORD_SET_DIALOG_ID))
    dispatch(setIsPasswordEnabled(true))
    dispatch(showNotification(intl.formatMessage(messages.account_password_enabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * disablePassword - Disables app-wide password protection.
 *
 * @param {string} password current password.
 * @returns {Function} Thunk
 */
export const disablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(password))
    await dispatch(waitForIpcEvent('deletePassword'))
    dispatch(setIsPasswordEnabled(false))
    dispatch(closeDialog(PASSWORD_PROMPT_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_disabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * requirePassword - Password protect routine. Should be placed before protected code.
 *
 * @param {string} password current password.
 * @returns {Promise} Promise that fulfills after login attempt (either successful or not)
 */
const requirePassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  const { password: hash } = await dispatch(waitForIpcEvent('getPassword'))
  const passwordHash = await sha256digest(password)
  // compare hash received from the main thread to a hash of a password provided
  if (hash === passwordHash) {
    return true
  }
  throw new Error(getIntl().formatMessage(messages.account_invalid_password))
}

/**
 * login - Perform account login.
 *
 * @param {string} password Password
 * @returns {Function} Thunk
 */
export const login = password => async dispatch => {
  try {
    dispatch({ type: LOGIN, password })
    await dispatch(requirePassword(password))
    dispatch({ type: LOGIN_SUCCESS })
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * clearLoginError - Clear login user error.
 *
 * @returns {Function} Thunk
 */
export const clearLoginError = () => (dispatch, getState) => {
  if (accountSelectors.loginError(getState())) {
    dispatch({
      type: 'LOGIN_CLEAR_ERROR',
    })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [INIT_ACCOUNT]: state => {
    state.isAccountLoading = true
  },
  [INIT_ACCOUNT_SUCCESS]: state => {
    state.isAccountLoading = false
    state.isAccountLoaded = true
    state.initAccountError = null
  },
  [INIT_ACCOUNT_FAILURE]: (state, { error }) => {
    state.isAccountLoading = false
    state.initAccountError = error
  },

  [LOGIN]: state => {
    state.isLoggingIn = true
  },
  [LOGIN_SUCCESS]: state => {
    state.isLoggingIn = false
    state.isLoggedIn = true
    state.loginError = null
  },
  [LOGIN_FAILURE]: (state, { error }) => {
    state.isLoggingIn = false
    state.loginError = error
  },
  [LOGIN_CLEAR_ERROR]: state => {
    state.loginError = null
  },
  [SET_IS_PASSWORD_ENABLED]: (state, { value }) => {
    state.isPasswordEnabled = value
  },
}

// ------------------------------------
// Reducer
// ------------------------------------

export default createReducer(initialState, ACTION_HANDLERS)
