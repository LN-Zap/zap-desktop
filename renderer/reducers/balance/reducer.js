import Coin from '@zap/utils/coin'
import createReducer from '@zap/utils/createReducer'
import { grpc } from 'workers'

import * as constants from './constants'

const { FETCH_BALANCE, FETCH_BALANCE_SUCCESS, FETCH_BALANCE_FAILURE } = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {boolean} isBalanceLoading Boolean indicating if balances are loading.
 * @property {string|null} walletBalance Total wallet balance.
 * @property {string|null} walletBalanceConfirmed Confirmed wallet balance.
 * @property {string|null} walletBalanceUnconfirmed Unconfirmed wallet balance.
 * @property {string|null} channelBalance Total channel balance.
 * @property {string|null} channelBalanceConfirmed Confirmed channel balance.
 * @property {string|null} channelBalancePending Pending channel balance.
 * @property {Error|null} fetchBalanceError Balance fetching error.
 */

/** @type {State} */
const initialState = {
  isBalanceLoading: false,
  walletBalance: null,
  walletBalanceConfirmed: null,
  walletBalanceUnconfirmed: null,
  channelBalance: null,
  channelBalanceConfirmed: null,
  channelBalancePending: null,
  fetchBalanceError: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * fetchBalance - Fetch balances.
 *
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const fetchBalance = () => async dispatch => {
  try {
    dispatch({ type: FETCH_BALANCE })
    const { walletBalance, channelBalance } = await grpc.services.Lightning.getBalance()
    dispatch({ type: FETCH_BALANCE_SUCCESS, walletBalance, channelBalance })
  } catch (error) {
    dispatch({ type: FETCH_BALANCE_FAILURE, error })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [FETCH_BALANCE]: state => {
    state.isBalanceLoading = true
  },
  [FETCH_BALANCE_SUCCESS]: (state, { walletBalance, channelBalance }) => {
    state.isBalanceLoading = false
    state.walletBalance = walletBalance.totalBalance
    state.walletBalanceConfirmed = walletBalance.confirmedBalance
    state.walletBalanceUnconfirmed = walletBalance.unconfirmedBalance
    state.channelBalance = Coin(channelBalance.balance)
      .add(Coin(channelBalance.pendingOpenBalance))
      .toString()
    state.channelBalanceConfirmed = channelBalance.balance
    state.channelBalancePending = channelBalance.pendingOpenBalance
  },
  [FETCH_BALANCE_FAILURE]: (state, { error }) => {
    state.isBalanceLoading = false
    state.fetchBalanceError = error
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
