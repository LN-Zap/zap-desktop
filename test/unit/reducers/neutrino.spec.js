import snapshotDiff from '../__helpers__/snapshotDiff'
import reducer, {
  START_NEUTRINO,
  START_NEUTRINO_SUCCESS,
  START_NEUTRINO_FAILURE,
  STOP_NEUTRINO,
  STOP_NEUTRINO_SUCCESS,
  STOP_NEUTRINO_FAILURE,
  RECEIVE_CURRENT_BLOCK_HEIGHT,
  RECEIVE_LND_BLOCK_HEIGHT,
  RECEIVE_LND_CFILTER_HEIGHT,
  RECEIVE_LND_RECOVERY_HEIGHT,
  SET_SYNC_STATUS_PENDING,
  SET_SYNC_STATUS_WAITING,
  SET_SYNC_STATUS_IN_PROGRESS,
  SET_SYNC_STATUS_COMPLETE,
  SET_GRPC_ACTIVE_INTERFACE,
  NEUTRINO_CRASHED,
  NEUTRINO_RESET,
} from 'reducers/neutrino'

describe('reducers', () => {
  describe('neutrinoReducer', () => {
    it('should handle initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot()
    })

    it('should handle START_NEUTRINO', () => {
      const action = {
        type: START_NEUTRINO,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle START_NEUTRINO_SUCCESS', () => {
      const action = {
        type: START_NEUTRINO_SUCCESS,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle START_NEUTRINO_FAILURE', () => {
      const action = {
        type: START_NEUTRINO_FAILURE,
        startNeutrinoError: 'Some error',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle STOP_NEUTRINO', () => {
      const action = {
        type: STOP_NEUTRINO,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle STOP_NEUTRINO_SUCCESS', () => {
      const action = {
        type: STOP_NEUTRINO_SUCCESS,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle STOP_NEUTRINO_FAILURE', () => {
      const action = {
        type: STOP_NEUTRINO_FAILURE,
        error: 'some error',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle RECEIVE_CURRENT_BLOCK_HEIGHT', () => {
      const action = {
        type: RECEIVE_CURRENT_BLOCK_HEIGHT,
        blockHeight: 12345,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle RECEIVE_LND_BLOCK_HEIGHT', () => {
      const action = {
        type: RECEIVE_LND_BLOCK_HEIGHT,
        data: [{ height: 12345 }],
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle RECEIVE_LND_CFILTER_HEIGHT', () => {
      const action = {
        type: RECEIVE_LND_CFILTER_HEIGHT,
        data: [{ height: 12345 }],
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle RECEIVE_LND_RECOVERY_HEIGHT', () => {
      const action = {
        type: RECEIVE_LND_RECOVERY_HEIGHT,
        data: [{ height: 12345 }],
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_SYNC_STATUS_PENDING', () => {
      const action = {
        type: SET_SYNC_STATUS_PENDING,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_SYNC_STATUS_WAITING', () => {
      const action = {
        type: SET_SYNC_STATUS_WAITING,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_SYNC_STATUS_IN_PROGRESS', () => {
      const action = {
        type: SET_SYNC_STATUS_IN_PROGRESS,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_SYNC_STATUS_COMPLETE', () => {
      const action = {
        type: SET_SYNC_STATUS_COMPLETE,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_GRPC_ACTIVE_INTERFACE', () => {
      const action = {
        type: SET_GRPC_ACTIVE_INTERFACE,
        grpcActiveInterface: 'lightning',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle NEUTRINO_CRASHED', () => {
      const action = {
        type: NEUTRINO_CRASHED,
        code: 1,
        signal: 'SIGTERM',
        lastError: 'Some error',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle NEUTRINO_RESET', () => {
      const action = {
        type: NEUTRINO_RESET,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })
  })
})
