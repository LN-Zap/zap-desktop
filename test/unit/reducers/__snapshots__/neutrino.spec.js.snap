// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers neutrinoReducer should handle NEUTRINO_CRASHED 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
-   "isNeutrinoCrashed": false,
+   "isNeutrinoCrashed": true,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
-   "neutrinoCrashCode": null,
-   "neutrinoCrashLastError": null,
-   "neutrinoCrashSignal": null,
+   "neutrinoCrashCode": 1,
+   "neutrinoCrashLastError": "Some error",
+   "neutrinoCrashSignal": "SIGTERM",
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle NEUTRINO_RESET 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`reducers neutrinoReducer should handle RECEIVE_CURRENT_BLOCK_HEIGHT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
-   "blockHeight": 0,
+   "blockHeight": 12345,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle RECEIVE_LND_BLOCK_HEIGHT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
-   "neutrinoBlockHeight": 0,
+   "neutrinoBlockHeight": 12345,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
-   "neutrinoFirstBlockHeight": 0,
+   "neutrinoFirstBlockHeight": 12345,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle RECEIVE_LND_CFILTER_HEIGHT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
-   "neutrinoCfilterHeight": 0,
+   "neutrinoCfilterHeight": 12345,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
-   "neutrinoFirstCfilterHeight": 0,
+   "neutrinoFirstCfilterHeight": 12345,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle RECEIVE_LND_RECOVERY_HEIGHT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
-   "neutrinoFirstRecoveryHeight": 0,
-   "neutrinoRecoveryHeight": 0,
+   "neutrinoFirstRecoveryHeight": 12345,
+   "neutrinoRecoveryHeight": 12345,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle SET_GRPC_ACTIVE_INTERFACE 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
-   "grpcActiveInterface": null,
+   "grpcActiveInterface": "lightning",
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle SET_SYNC_STATUS_COMPLETE 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
-   "syncStatus": "pending",
+   "syncStatus": "complete",
  }
`;

exports[`reducers neutrinoReducer should handle SET_SYNC_STATUS_IN_PROGRESS 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
-   "syncStatus": "pending",
+   "syncStatus": "in-progress",
  }
`;

exports[`reducers neutrinoReducer should handle SET_SYNC_STATUS_PENDING 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`reducers neutrinoReducer should handle SET_SYNC_STATUS_WAITING 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
-   "syncStatus": "pending",
+   "syncStatus": "waiting",
  }
`;

exports[`reducers neutrinoReducer should handle START_NEUTRINO 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
-   "isStartingNeutrino": false,
+   "isStartingNeutrino": true,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle START_NEUTRINO_FAILURE 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
-   "startNeutrinoError": null,
+   "startNeutrinoError": "Some error",
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle START_NEUTRINO_SUCCESS 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
-   "isNeutrinoRunning": false,
+   "isNeutrinoRunning": true,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle STOP_NEUTRINO 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
-   "isStoppingNeutrino": false,
+   "isStoppingNeutrino": true,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
    "stopNeutrinoError": null,
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle STOP_NEUTRINO_FAILURE 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "blockHeight": 0,
    "grpcActiveInterface": null,
    "isNeutrinoCrashed": false,
    "isNeutrinoRunning": false,
    "isStartingNeutrino": false,
    "isStoppingNeutrino": false,
    "neutrinoBlockHeight": 0,
    "neutrinoCfilterHeight": 0,
    "neutrinoCrashCode": null,
    "neutrinoCrashLastError": null,
    "neutrinoCrashSignal": null,
    "neutrinoFirstBlockHeight": 0,
    "neutrinoFirstCfilterHeight": 0,
    "neutrinoFirstRecoveryHeight": 0,
    "neutrinoRecoveryHeight": 0,
    "startNeutrinoError": null,
-   "stopNeutrinoError": null,
+   "stopNeutrinoError": "some error",
    "syncStatus": "pending",
  }
`;

exports[`reducers neutrinoReducer should handle STOP_NEUTRINO_SUCCESS 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`reducers neutrinoReducer should handle initial state 1`] = `
Object {
  "blockHeight": 0,
  "grpcActiveInterface": null,
  "isNeutrinoCrashed": false,
  "isNeutrinoRunning": false,
  "isStartingNeutrino": false,
  "isStoppingNeutrino": false,
  "neutrinoBlockHeight": 0,
  "neutrinoCfilterHeight": 0,
  "neutrinoCrashCode": null,
  "neutrinoCrashLastError": null,
  "neutrinoCrashSignal": null,
  "neutrinoFirstBlockHeight": 0,
  "neutrinoFirstCfilterHeight": 0,
  "neutrinoFirstRecoveryHeight": 0,
  "neutrinoRecoveryHeight": 0,
  "startNeutrinoError": null,
  "stopNeutrinoError": null,
  "syncStatus": "pending",
}
`;
